library/alloc/src/collections/binary_heap.rs

```diff
@@ -1220,6 +1220,30 @@ impl<T> BinaryHeap<T> {
     pub fn clear(&mut self) {
         self.drain();
     }
+
+    /// Creates an iterator which uses a closure to determine if an element should be removed from the underlying vec.
+    /// If the closure returns true, then the element is removed and yielded. If the closure returns false, the element will remain in the binary heap and will not be yielded by the iterator.
+    /// # Examples
+    /// Splitting the binary heap into evens and odds, reusing the original allocation:
+    ///
+    /// ```
+    /// #![feature(binary_heap_drain_filter)]
+    /// use std::collections::BinaryHeap;
+    /// let mut a = BinaryHeap::from(vec![1, 2, 3, 4, 5, 6, 8, 9, 11, 13, 14, 15]);
+    /// let mut evens = a.drain_filter(|x| *x % 2 == 0).collect::<Vec<_>>();
+    /// evens.sort();
+    /// let odds = a.into_sorted_vec();
+    ///
+    /// assert_eq!(evens, vec![2, 4, 6, 8, 14]);
+    /// assert_eq!(odds, vec![1, 3, 5, 9, 11, 13, 15]);
+    /// ```
+    #[unstable(feature = "binary_heap_drain_filter", reason = "recently added", issue = "42849")]
+    pub fn drain_filter<F>(&mut self, filter: F) -> DrainFilter<'_, T, F>
+    where
+        F: FnMut(&mut T) -> bool,
```

Comment:
Taking a `&mut` closure would allow the iterator to modify the items and violate heap order.

Reply:
Should this be changed to `& T`? or should the closure take `PeekMut`? that way if the item is mutated, the binary heap will sift down as per the `PeekMut` implementation.